akd_local_auditor/src/storage/s3.rs

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under both the MIT license found in the
+// LICENSE-MIT file in the root directory of this source tree and the Apache
+// License, Version 2.0 found in the LICENSE-APACHE file in the root directory
+// of this source tree.
+
+//! This module comprises S3 bucket READ ONLY access to download and parse
+//! Audit Proofs
+
+use anyhow::{bail, Result};
+use async_trait::async_trait;
+use aws_config::RetryConfig;
+use aws_sdk_s3 as s3;
+use clap::Args;
+use log::{debug, error};
+use std::convert::TryInto;
+use std::sync::Arc;
+use tokio::sync::RwLock;
+
+const MIN_BUCKET_CHARS: usize = 3;
+const MAX_BUCKET_CHARS: usize = 63;
+const ALLOWED_BUCKET_CHARS: [char; 38] = [
+    'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's',
+    't', 'u', 'v', 'w', 'x', 'y', 'z', '.', '-', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0',
+];
+
+fn is_bucket_name_valid(s: &str) -> Result<String, String> {
+    let str = s.to_string();
+    if str.len() >= MIN_BUCKET_CHARS && str.len() <= MAX_BUCKET_CHARS {
+        for c in str.chars() {
+            if !ALLOWED_BUCKET_CHARS.iter().any(|v| c == *v) {
+                return Err(format!("Character '{}' is not allowed in bucket name. Bucket names must contain lower-case letters, numbers, '-', and '.' only.", c));
+            }
+        }
+        Ok(str)
+    } else {
+        Err(format!(
+            "Bucket name must be between [{}, {}] characters in length. Gave {}",
+            MIN_BUCKET_CHARS,
+            MAX_BUCKET_CHARS,
+            str.len()
+        ))
+    }
```

Comment:
Just a suggestion, but you can invert the if and return the `Err` to avoid a level a nesting with the inner for-loop. Additionally, I think the inner check for validating all characters by doing something like [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=be8299413bfb44c6bb22bcda7ba02f85).

Reply:
Ah a remnant of my functional programming days. early exit's are "frowned upon" often, as it makes tracing exits a lot more complicated sometimes. But we aren't writing functional code here lol

Reply:
Haha no worries! If you prefer this style, that's fine with me. This was mostly a nitpick type comment that I didn't explicitly categorize as such!